### DISCUSSION TITLE
Where should I put override "to_hstring" function of enum in WinUI Desktop App with C++/WinRT?

### DISCUSSION THREAD

**Discussion by sxlllslgh** [microsoft/WindowsAppSDK#993] [Q&A] [answered]
I want to convert enum type to string to show in a ListView:
```xml
<ListView x:Name="MyList" ItemTemplate="{StaticResource MyListViewTemplate}" ItemsSource="{x:Bind MyListViewModel.List, Mode=OneWay}" />
```
The **_List_** function of **_MyListViewModel_** returns an **_ObservableVector<MyItem>_**, and class **_MyItem_** contains an enum type defined in IDL:
```
enum MyEnum {
    Monday,
    Tuesday
}

runtimeclass MyItem {
    MyEnum MyEnum;
}
```
```cpp
namespace winrt::MyApp::implementation {
    struct MyItem : MyItemT<MyItem> {
    private:
        MyEnum myEnum;
    public:
        MyEnum MyEnum();
        void MyEnum(MyEnum const& newMyEnum);
    };
}
```
Then I override **_to_hstring_** function of **_MyEnum_** according the [UWP sample](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/move-to-winrt-from-cx#tostring):
```cpp
namespace winrt {
    hstring to_hstring(MyEnum myEnum) {
        switch (myEnum) {
            case StatusEnum::Monday: return L"Start a new week";
            case StatusEnum::Tuesday: return L"Balabala";
            default: return L"Error";
        }
    }
}
```
But wherever I put this function, such as App.h, App.cpp, MainWindow.h, MainWindow.cpp... I would always get errors like:
```bash
MyItemViewModel.obj : error LNK2005: "struct winrt::hstring __cdecl winrt::to_hstring(enum winrt::MyApp::MyEnum)" (?to_hstring@winrt@@YA?AUhstring@1@W4MyEnum@MyApp@1@@Z) has been defined in MyItem.obj
```
It seems C++/WinRT would always generate default to_hstring function of enum class, so how to override it to show in XAML UI?

**Comment by jonwis** [comment, marked as answer]
You probably want a data binding converter - https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.data.binding.converter?view=winui-3.0#Microsoft_UI_Xaml_Data_Binding_Converter but @jevansaks  or @MikeHillberg can tell you more.

**Reply by jevansaks** [reply]
Since you're already using x:Bind, you can use functions in your x:Bind path. https://docs.microsoft.com/en-us/windows/uwp/xaml-platform/x-bind-markup-extension#functions-in-binding-paths

But this error seems different -- I suspect you've defined a function with a body in a header that's included into two compilation units and then the linker gets upset because there's two definitions.

Try adding "inline" specifier on your `to_hstring` function.

**Reply by sxlllslgh** [reply]
Thank you for your help! **`inline`** function works now! Later I would try to use custom converter.
